### PR TITLE
Pro 5604: translation during localization backend

### PR DESCRIPTION
**Pull request by myovchev**
## Summary

Implement the backend methods, handlers and REST API. 

## What are the specific steps to test this change?

*For example:*
> 1. Run the website and log in as an admin
> 2. Open a piece manager modal and select several pieces
> 3. Click the "Archive" button on the top left of the manager and confirm that it should proceed
> 4. Check that all pieces have been archived properly

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [ ] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [ ] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [ ] The changelog is updated
- [ ] Related documentation has been updated
- [ ] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: modules/@apostrophecms/translation/index.js [hunk 1]
```diff
@@ -1,17 +1,173 @@
...
+    return {
+      '@apostrophecms/doc-type:beforeLocalize': {
+        // Translate the document using the first available provider.
+        async translate(req, doc, source, target, options) {
+          if (!req.query.aposTranslateTargets?.includes(target)) {
+            return;
+          }
+          // We don't support multiple providers yet, so just use the first one
+          // when no provider is specified.
+          const name = self.getProvider(req.query.aposTranslateProvider)?.name;
```
Comment (ValJed) [inline modules/@apostrophecms/translation/index.js:22]: Shouldn't we validate query params using `launder`?
Even if there are passed by ourself in the vue app.

File: modules/@apostrophecms/translation/index.js [hunk 2]
```diff
@@ -1,17 +1,173 @@
...
+            req.t('apostrophe:automaticTranslationLngCheckNoProvider')
+          );
+        }
+
+        return module.getSupportedLanguages(
+          req,
+          name,
+          req.query.source ?? [],
+          req.query.target ?? []
+        );
```
Comment (ValJed) [inline modules/@apostrophecms/translation/index.js:96]: [optional]
For readability I would pass req as a first argument and the rest in  an object
```javascript
      return module.getSupportedLanguages(
          req,
          {
            name,
            source: req.query.source ?? [],
            target: req.query.target ?? []
          }
        );
```

**Review by ValJed — APPROVED** [review]
Approving, still few questions / propositions

File: modules/@apostrophecms/doc/index.js [hunk 3]
```diff
@@ -937,6 +937,252 @@ module.exports = {
...
+      // A cleanup is performed to remove empty meta properties on each call.
+      removeMeta(doc, namespace, ...pathArgsWithKey) {
+        if (!doc || !namespace) {
+          throw self.apos.error('invalid', 'Document and namespace are required.', {
+            cause: 'invalidArguments'
+          });
+        }
+
+        const pathArgs = [ ...pathArgsWithKey ];
+        const key = pathArgs.pop();
```
Comment (ValJed) [inline modules/@apostrophecms/doc/index.js:1063]: Same as above

File: modules/@apostrophecms/doc/index.js [hunk 4]
```diff
@@ -937,6 +937,252 @@ module.exports = {
           return self.db.insertOne(self.apos.util.clonePermanent(doc));
         });
       },
+      // Set meta data for a given field, that will be live under `aposMeta`
+      // doc property. It returns the path to the meta property withouth the
+      // key. See `getMetaPath` method for more information.
+      //
+      // Signature:
```
Comment (ValJed) [inline modules/@apostrophecms/doc/index.js:944]: Why not using jsdoc like you did in vue mixin?

**Review by ValJed — APPROVED** [review]
questions / propositions

File: modules/@apostrophecms/doc-type/index.js [hunk 5]
```diff
@@ -1169,8 +1186,12 @@ module.exports = {
             aposLocale: `${toLocale}:draft`,
             metaType: 'doc'
           };
+          await self.emit('beforeLocalize', req, update, eventOptions);
           result = await actionModule.update(toReq, update);
         }
+
+        await self.emit('afterLocalize', req, draft, result, eventOptions);
```
[the remark this answers is not in the document]
Comment (myovchev) [reply modules/@apostrophecms/doc-type/index.js:1193]: `result` is the inserted/updated document, draft is the original.

File: modules/@apostrophecms/translation/index.js [hunk 6]
```diff
@@ -4,14 +4,169 @@ module.exports = {
...
+            { cause: 'invalidArguments' }
+          );
+        }
+        if (typeof aposModule.translate !== 'function') {
+          throw new Error(
+            `The translation provider module "${aposModule.__meta.name}" has to implement "translate" method.`,
+            { cause: 'interfaceNotImplemented' }
+          );
+        }
+        if (typeof aposModule.getSupportedLanguages !== 'function') {
```
[the remark this answers is not in the document]
Comment (myovchev) [reply modules/@apostrophecms/translation/index.js:152]: If you pass empty source and/or target it returns all languages. So it's more "get", source and target are  "filters" that we respect.

File: modules/@apostrophecms/translation/index.js (hunk 2 above)
Comment (myovchev) [reply modules/@apostrophecms/translation/index.js:96]: Also this one should be "laundered" I guess.

**Pull request merged**